src/lib/comms/sol-coap.c

```diff
@@ -969,47 +969,66 @@ find_context(struct sol_coap_server *server, const struct sol_coap_resource *res
     return NULL;
 }
 
-SOL_API int
-sol_coap_notify(struct sol_coap_server *server,
-    struct sol_coap_resource *resource, struct sol_coap_packet *pkt)
+static int
+sol_coap_notify_full(struct sol_coap_server *server,
+    struct sol_coap_resource *resource, struct sol_coap_packet *pkt,
+    int (*cb)(void *cb_data, struct sol_coap_server *server,
+    struct sol_coap_resource *resource, struct sol_network_link_addr *addr,
+    struct sol_coap_packet **pkt),
+    const void *cb_data)
 {
     struct resource_observer *o;
     struct resource_context *c;
     struct sol_coap_packet *header;
     uint16_t i;
     int r = -EINVAL;
 
-    SOL_NULL_CHECK(pkt, -EINVAL);
-    SOL_NULL_CHECK_GOTO(server, error_pkt);
-    SOL_NULL_CHECK_GOTO(resource, error_pkt);
+    SOL_NULL_CHECK_GOTO(server, err_pkt);
+    SOL_NULL_CHECK_GOTO(resource, err_pkt);
 
-    COAP_RESOURCE_CHECK_API_GOTO(error_pkt);
+    COAP_RESOURCE_CHECK_API_GOTO(err_pkt);
 
     c = find_context(server, resource);
     if (!c) {
-        SOL_WRN("Context not found for expecified resource");
+        SOL_WRN("Context not found for specified resource");
         r = -ENOENT;
-        goto error_pkt;
+        goto err_pkt;
     }
 
     r = 0;
     SOL_PTR_VECTOR_FOREACH_IDX (&c->observers, o, i) {
         uint8_t type, code;
 
+        if (cb) {
+            r = cb((void *)cb_data, server, resource, &o->cliaddr, &pkt);
+            if (r < 0 && r != -EPERM) {
+                SOL_WRN("Error creating notification packet. Reason: %d", r);
+
+                goto err_pkt;
+            } else if (r == -EPERM) {
+                SOL_BUFFER_DECLARE_STATIC(addr, SOL_NETWORK_INET_ADDR_STR_LEN);
+
+                sol_network_link_addr_to_str(&o->cliaddr, &addr);
+                SOL_WRN("Observer at %.*s is not authorized for CoAP Resource %p",
+                    SOL_STR_SLICE_PRINT(sol_buffer_get_slice(&addr)), resource);
+                continue;
+            }
+        }
+
         header = sol_coap_packet_new(NULL);
         if (!header) {
             r = -ENOMEM;
-            goto error_pkt;
+            goto err_pkt;
         }
 
         sol_coap_header_get_code(pkt, &code);
         r = sol_coap_header_set_code(header, code);
-        SOL_INT_CHECK_GOTO(r, < 0, err);
+        SOL_INT_CHECK_GOTO(r, < 0, err_header);
         sol_coap_header_get_type(pkt, &type);
         r = sol_coap_header_set_type(header, type);
-        SOL_INT_CHECK_GOTO(r, < 0, err);
+        SOL_INT_CHECK_GOTO(r, < 0, err_header);
         r = sol_coap_header_set_token(header, o->token, o->tkl);
-        SOL_INT_CHECK_GOTO(r, < 0, err);
+        SOL_INT_CHECK_GOTO(r, < 0, err_header);
```

Comment:
This will be only necessary is `!cb`, otherwise the packet returned by `cb` will already contains the correct token and stuff. 
The user is supposed to use `sol_coap_notification_packet_new()`